File: Java/Algorithms/HeapSort.java
package Algorithms;

public class HeapSort {
	
	public static void sort(int[] arr) {
		int size = arr.length;
		// Heapify
		for(int i = size / 2 - 1; i >= 0; i--) {
			heapify(arr, size, i);
		}
		
		int i = size - 1;
		while (i >= 1) {
			// Swap the first with arr[i]. 
			swap(arr, 0, i);
			// Heapify the array again.
			heapify(arr, i, 0);
			// Then decrement i.
			i--;
		}
	}
	
	private static void heapify(int[] arr, int size, int i) {
		
		int largest = i;
		// i-th element's left child.
		int leftLeaf = 2*i + 1;
		// i-th element's right child.
		int rightLeaf = 2*i + 2;
		
		// If the left child is larger than the current largest.
		if (leftLeaf < size && arr[leftLeaf] > arr[largest]) {
			largest = leftLeaf;
		}
		// If the right child is larger than the current largest.
		if (rightLeaf < size && arr[rightLeaf] > arr[largest]) {
			largest = rightLeaf;
		}
		
		// If the largest of the two is not the original largest
		if (largest != i) {
			// Swap i and the largest.
			swap(arr, i, largest);
			// Heapify the sub-tree. 
			heapify(arr, size, largest); 
		}
	}
	
	private static void swap(int[] arr, int a, int b) {
		int tmp = arr[a];
		arr[a]  = arr[b];
		arr[b] = tmp;
	}

}

File: TypeScript/Algorithms/HeapSort.ts
class HeapSort
{
    public sort(arr: number[])
    {
        const size = arr.length;

        for(let i = Math.floor(size / 2) - 1; i >= 0; i--)
        {
            this.heapify(arr, size, i);
        }

        let j = size - 1;
        while(j >= 1)
        {
            this.swap(arr, 0, j);
            this.heapify(arr, j, 0);
            j--;
        }
    }

    private heapify(arr: number[], size: number, i: number)
    {
        let largest = i;

        let leftLeaf = 2 * i + 1;
        let rightLeaf = 2 * i + 2;

        // If the left child is larger than the current largest.
		if (leftLeaf < size && arr[leftLeaf] > arr[largest]) 
        {
			largest = leftLeaf;
		}
		// If the right child is larger than the current largest.
		if (rightLeaf < size && arr[rightLeaf] > arr[largest]) 
        {
			largest = rightLeaf;
		}
		
		// If the largest of the two is not the original largest
		if (largest != i) 
        {
			// Swap i and the largest.
			this.swap(arr, i, largest);
			// Heapify the sub-tree. 
			this.heapify(arr, size, largest); 
		}
    }

    private swap(arr: number[], a: number, b: number)
    {
        const tmp = arr[a];
        arr[a] = arr[b];
        arr[b] = tmp;
    }
}